presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/TemporalFunctionFactory.java

```diff
@@ -0,0 +1,132 @@
+import com.facebook.presto.spi.type.SqlDate;
+import com.facebook.presto.spi.type.SqlTimestamp;
+import com.facebook.presto.spi.type.Type;
+import org.joda.time.DateTimeZone;
+
+import javax.inject.Inject;
+
+import java.time.Duration;
+
+import static com.facebook.presto.spi.type.DateType.DATE;
+import static com.facebook.presto.spi.type.TimestampType.TIMESTAMP;
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class TemporalFunctionFactory
+{
+    private final DateTimeZone shardDayBoundaryTimeZone;
+
+    @Inject
+    public TemporalFunctionFactory(StorageManagerConfig config)
+    {
+        this.shardDayBoundaryTimeZone = requireNonNull(config.getShardDayBoundaryTimeZone(), "shardDayBoundaryTimeZone is null");
+    }
+
+    public TemporalFunction createTemporalFunction(Type type)
+    {
+        if (type == DATE) {
+            return new DateTemporalFunction();
+        }
+        else if (type.equals(TIMESTAMP)) {
+            return new TimestampTemporalFunction(shardDayBoundaryTimeZone);
+        }
+
+        throw new IllegalArgumentException("Wrong type for temporal column: " + type);
+    }
+
+    public static class DateTemporalFunction
+            implements TemporalFunction
+
+    {
+        @Override
+        public int getDay(ConnectorSession session, Block block, int position)
+        {
+            return toIntExact(((SqlDate) DATE.getObjectValue(session, block, position)).getDays());
+        }
+
+        @Override
+        public int getDay(Object min, Object max)
+        {
+            return ((Integer) min).intValue();
+        }
+
+        @Override
+        public Type getType()
+        {
+            return DATE;
+        }
+    }
+
+    public static class TimestampTemporalFunction
+            implements TemporalFunction
+    {
+        DateTimeZone timeZone;
+
+        public TimestampTemporalFunction(DateTimeZone timeZone)
+        {
+            this.timeZone = requireNonNull(timeZone, "timeZone is null");
+        }
+
+        @Override
+        public int getDay(ConnectorSession session, Block block, int position)
+        {
+            SqlTimestamp timestamp = (SqlTimestamp) TIMESTAMP.getObjectValue(session, block, position);
+            long millis = timestamp.getMillisUtc();
+            long local = timeZone.convertUTCToLocal(millis);
+            long days = MILLISECONDS.toDays(local);
+            return toIntExact(days);
+        }
+
+        @Override
+        public int getDay(Object min, Object max)
+        {
+            Long minValue = timeZone.convertUTCToLocal((Long) min);
+            Long maxValue = timeZone.convertUTCToLocal((Long) max);
+            return determineDay(minValue, maxValue);
+        }
+
+        @Override
+        public Type getType()
+        {
+            return TIMESTAMP;
+        }
+
+        private int determineDay(long rangeStart, long rangeEnd)
```

Comment:
Ideally, we should keep this logic in `ShardOrganizerUtils` and only extract the part related to date boundary/timezone.